hail/python/hailtop/aiotools/fs/copier.py

```diff
@@ -1,446 +1,20 @@
-    # >>> import os
-    # >>> entries = os.scandir('.')
-    # >>> type(entries)
-    # <class 'posix.ScandirIterator'>
-    # >>> import posix
-    # >>> posix.ScandirIterator
-    # Traceback (most recent call last):
-    #   File "<stdin>", line 1, in <module>
-    # AttributeError: module 'posix' has no attribute 'ScandirIterator'
-    async def _listfiles_recursive(self, url: str, entries) -> AsyncIterator[FileListEntry]:
-        async for file in self._listfiles_flat(url, entries):
-            if await file.is_file():
-                yield file
-            else:
-                new_url = await file.url()
-                new_path = self._get_path(new_url)
-                new_entries = await blocking_to_async(self._thread_pool, os.scandir, new_path)
-                async for subfile in self._listfiles_recursive(new_url, new_entries):
-                    yield subfile
-
-    async def _listfiles_flat(self, url: str, entries) -> AsyncIterator[FileListEntry]:
-        with entries:
-            for entry in entries:
-                yield LocalFileListEntry(self._thread_pool, url, entry)
-
-    async def listfiles(self, url: str, recursive: bool = False) -> AsyncIterator[FileListEntry]:
-        path = self._get_path(url)
-        entries = await blocking_to_async(self._thread_pool, os.scandir, path)
-        if recursive:
-            return self._listfiles_recursive(url, entries)
-        return self._listfiles_flat(url, entries)
-
-    async def staturl(self, url: str) -> str:
-        path = self._get_path(url)
-        stat_result = await blocking_to_async(self._thread_pool, os.stat, path)
-        if stat.S_ISDIR(stat_result.st_mode):
-            return AsyncFS.DIR
-        return AsyncFS.FILE
-
-    async def mkdir(self, url: str) -> None:
-        path = self._get_path(url)
-        await blocking_to_async(self._thread_pool, os.mkdir, path)
-
-    async def makedirs(self, url: str, exist_ok: bool = False) -> None:
-        path = self._get_path(url)
-        await blocking_to_async(self._thread_pool, os.makedirs, path, exist_ok=exist_ok)
-
-    async def isfile(self, url: str) -> bool:
-        path = self._get_path(url)
-        return await blocking_to_async(self._thread_pool, os.path.isfile, path)
-
-    async def isdir(self, url: str) -> bool:
-        path = self._get_path(url)
-        return await blocking_to_async(self._thread_pool, os.path.isdir, path)
-
-    async def remove(self, url: str) -> None:
-        path = self._get_path(url)
-        return await blocking_to_async(self._thread_pool, os.remove, path)
-
-    async def rmtree(self, sema: Optional[asyncio.Semaphore], url: str) -> None:
-        path = self._get_path(url)
-
-        def f():
-            try:
-                shutil.rmtree(path)
-            except FileNotFoundError:
-                # no error if path does not exist
-                pass
-
-        await blocking_to_async(self._thread_pool, f)
+if TYPE_CHECKING:
```

Comment:
We should treat `if TYPE_CHECKING` as a code smell. This code can be restructured to avoid the circularity: https://github.com/jigold/hail/compare/lazy-router-pt2...danking:dk-lazy-router-pt2 (I also included the copy part fix)

Reply:
Erm, I didn't mention the copy part fix before. The copier shouldn't directly depend on RouterAsyncFS. We might want to use it to copy between the same file system. We can avoid this by making `copy_part_size` a function that takes the URL. From the URL it determines the scheme and returns an appropriate copy part size.